constellations/orion/src/controllers/classes_by_semesters_controller.go

```diff
@@ -0,0 +1,125 @@
+	semesterMap := make(map[string]domains.Semester)
+	classMap := make(map[string]domains.Class)
+
+	for i := 0; i < len(programs); i++ {
+		programMap[programs[i].ProgramId] = programs[i]
+	}
+	for i := 0; i < len(semesters); i++ {
+		semesterMap[semesters[i].SemesterId] = semesters[i]
+	}
+	for i := 0; i < len(classes); i++ {
+		classMap[classes[i].ClassId] = classes[i]
+	}
+
+	// Semester to class list map (maps semesterId to Class object)
+	semesterClassMap := make(map[string][]domains.Class)
+
+	// Loop over semesterIds
+	for i := 0; i < len(semesters); i++ {
+		// Create Semester Object
+		semesterObj := semesters[i]
+		var programClassStruct domains.ProgramClass
+		var programObj domains.Program
+
+		// Find classes that have the same semesterId and append to semesterClassMap
+		classSlice := make([]domains.Class, 0)
+		for i := 0; i < len(classes); i++ {
+			if classes[i].SemesterId == semesterObj.SemesterId {
+				classSlice = append(classSlice, classes[i])
+			}
+		}
+		// List of classes in specific semester
+		semesterClassMap[semesterObj.SemesterId] = classSlice
+
+		// Create list of programIds in specific semester
+		var programList []string
+		for i := 0; i < len(classSlice); i++ {
+			if Find(programList, classSlice[i].ProgramId) != -1 {
+				continue
+			} else {
+				programList = append(programList, classSlice[i].ProgramId)
+			}
+		}
+
+		// Find all classes in each program
+		finalClassList := make([]domains.Class, 0)
+		listProgramClass := make([]domains.ProgramClass, 0)
+		for i := 0; i < len(programList); i++ {
+			for j := 0; j < len(classSlice); j++ {
+				if programList[i] == classSlice[j].ProgramId {
+					finalClassList = append(finalClassList, classSlice[j])
+				}
+			}
+
+			// Check if programId exists inside programMap
+			if value, ok := programMap[programList[i]]; ok {
+				programObj = value
+			} else {
+				c.Error(err)
+				c.String(http.StatusInternalServerError, err.Error())
+			}
+
+			// Make ProgramClass struct
+			programClassStruct = updateProgramClass(programObj, finalClassList)
+			// Append to list of ProgramClasses
+			listProgramClass = append(listProgramClass, programClassStruct)
+			finalClassList = nil
+		}
+
+		// Make ProgramClassesBySemester struct
+		classSemester = updateProgramClassesBySemester(semesters[i], listProgramClass)
+		listResults = append(listResults, classSemester)
+	}
+	if err != nil {
+		c.Error(err)
+		c.String(http.StatusNotFound, err.Error())
+	} else {
+		c.JSON(http.StatusOK, &listResults)
+	}
+}
+
+func updateProgramClass(programObj domains.Program, classes []domains.Class) domains.ProgramClass {
+	return domains.ProgramClass{
+		ProgramObj: programObj,
+		Classes:    classes,
+	}
+}
+
+func updateProgramClassesBySemester(semester domains.Semester, programClasses []domains.ProgramClass) domains.ProgramClassesBySemester {
+	return domains.ProgramClassesBySemester{
+		Semester:       semester,
+		ProgramClasses: programClasses,
+	}
+}
```

Comment:
I think these are nice tries, but I'm hoping these functions have more logic in them. These just create new objects which would probably help with shorthand. But I'm hoping more for functions with like for-loops in them or complicated map-creating in them, you know?